### PR TITLE
ad9361_baseband_auto_rate: Fix low rate tx filter gain

### PR DESCRIPTION
**Pull request by Wramberg**
We are using the AD9361 for transmission of a DVB-S2 signal with symbol rates between 1MBaud to 50 Mbaud. We use the `ad9361_set_bb_rate()` function for compensation in the TX path. We observe no impact on output signal when running with sample-rates above 20 MSamples/s. When running with sample-rates lesser than or equal to 20 Msamples/s we observe a 6 dB power drop when switching between the TX FIR filter being disabled and enabled (power drop occurs when enabling the filter). The power drop seems to be caused by the interpolation being doubled without any gain compensation in the filter.

This PR contains a straight forward fix by adjusting the TX filter gain by a factor of two which solves the issue of a power drop. Consider this an FYI about this issue with a naive approach for fixing it. I would expect the FIR filters to be gain adjusted such that the output power is not affected like this.

Edit: Assuming the dynamic range of the in- and output of the FIR filter is the same no overflow should occur as far as I can see.

Output power before the fix was applied
outpowerbeforefix

Output power after the fix was applied - no other setting changed!
outpowerafterefix

### CODE REVIEW

File: ad9361_baseband_auto_rate.c [hunk 1]
```diff
@@ -132,7 +144,7 @@ int ad9361_set_bb_rate(struct iio_device *dev, unsigned long rate)
 	len += snprintf(buf + len, FIR_BUF_SIZE - len, "TX 3 GAIN 0 INT %d\n", dec);
 
 	for (i = 0; i < taps; i++)
-		len += snprintf(buf + len, FIR_BUF_SIZE - len, "%d,%d\n", fir[i], fir[i]);
+		len += snprintf(buf + len, FIR_BUF_SIZE - len, "%d,%d\n", fir_tx[i], fir_rx[i]);
```
Comment (Wramberg) [inline ad9361_baseband_auto_rate.c:147]: I suppose we could also add a gain factor in the if-else construct where the filters are chosen and simply apply that gain factor here.

**Comment by tfcollins** [comment]
Scaling the taps in this way will cause saturation in the TFIR if you are not backed off. A simple test you can run is to send a full-scale waveform through with this filter loaded and read out the register 0x5E. D1 will tell you if the TFIR is saturating, which is happening for me with this change when transmitting a tone.

This might be fine for the DVB-S2 case but not for all signals which auto_baseband_rate tries to be.

-Travis

**Comment by Wramberg** [comment]
I think I am missing some vital information here so let me start from the top.

I hope we agree that the existing behavior is somewhat odd. Ideally:

1. The TX output power should not be affected by using the FIR filter provided by this lib right ?
2. Or, at the very least, it should be consistent across samples-rates so we do not observe a sudden 6 dB drop when going from 21 MHz to 19 MHz sample rate.

It seems to me that the `fir_128_2` coefficients have been scaled to achieve point 1. At least from my experiments with the spectrum analyzer there is no difference between having the FIR enabled or disabled when the `fir_128_2` coefficients are loaded.

Now I have had trouble finding a detailed description of the TX FIR filter design but here is what I know
- The coefficients are stored in 16-bit twos complement format [AD9361 Ref. Manual UG-570]
- The filter input is 12-bit twos complement format [AD9361 Ref. Manual UG-570]
- The interpolation is done before the filtering [Some ADI forum post]

And here is what I have assumed from inspecting in and output of the filter, please let me know if this is wrong
- Filter operations are done in 16 bit width
- The 12-bit input is left shifted with `0`s as LSBs to match coefficient width
- The output of the filter is truncated back 16 or 12 bit width. For simplicity lets assume 16 bit. This means the 15 LSBs (and a redundant sign bit) are removed at the filter output since that is 32 bit after multiplication.
- Interpolation is done by inserting `0`s

The above assumptions fit with what I have observed in my experiments. If we consider the `fir_128_2` coefficients in Q.15 format and convert back to floating point by multiplying them with 2^-15 we get the following frequency response.

filter_freq_resp

No power loss occurs for the `fir_128_2` filter as it has 6dB gain (factor of 2) which compensates for the interpolation factor of 2. The amplitude of the in- and output is thus maintained.  The `fir_128_4` however uses an interpolation factor of 4. As it has the same gain the output amplitude is half of the input amplitude thus causing a loss in output power of 6dB power.

I am uncertain if you agree with the above but I hope you do and your concern is simply with the scaling versus overflow. In relation to overflow I assume you are talking about overflow at the output and that the filter can handle internal overflow (2s complement wraparound) ?

If yes I do not understand your statement that `This might be fine for the DVB-S2 case but not for all signals which auto_baseband_rate tries to be.`  The existing `fir_128_4`  does not seem to be scaled to avoid overflow for all signals ? At least if my assumptions are correct you can quite easily cause overflow with a tone in that filter as well. If this is not the case I have clearly misunderstood something fundamental and would appreciate if you had the time to comment on my assumptions and elaborate on the inner workings.

**Comment by Wramberg** [comment]
I think I am missing some vital information here so let me start from the top.

I hope we agree that the existing behavior is somewhat odd. Ideally:

1. The TX output power should not be affected by using the FIR filter provided by this lib right ?
2. Or, at the very least, it should be consistent across samples-rates so we do not observe a sudden 6 dB drop when going from 21 MHz to 19 MHz sample rate.

It seems to me that the `fir_128_2` coefficients have been scaled to achieve point 1. At least from my experiments with the spectrum analyzer there is no difference between having the FIR enabled or disabled when the `fir_128_2` coefficients are loaded.

Now I have had trouble finding a detailed description of the TX FIR filter design but here is what I know
- The coefficients are stored in 16-bit twos complement format [AD9361 Ref. Manual UG-570]
- The filter input is 12-bit twos complement format [AD9361 Ref. Manual UG-570]
- The interpolation is done before the filtering [Some ADI forum post]

And here is what I have assumed from inspecting in and output of the filter, please let me know if this is wrong
- Filter operations are done in 16 bit width
- The 12-bit input is left shifted with `0`s as LSBs to match coefficient width
- The output of the filter is truncated back 16 or 12 bit width. For simplicity lets assume 16 bit. This means the 15 LSBs (and a redundant sign bit) are removed at the filter output since that is 32 bit after multiplication.
- Interpolation is done by inserting `0`s

The above assumptions fit with what I have observed in my experiments. If we consider the `fir_128_2` coefficients in Q.15 format and convert back to floating point by multiplying them with 2^-15 we get the following frequency response.

filter_freq_resp

No power loss occurs for the `fir_128_2` filter as it has 6dB gain (factor of 2) which compensates for the interpolation factor of 2. The amplitude of the in- and output is thus maintained.  The `fir_128_4` however uses an interpolation factor of 4. As it has the same gain the output amplitude is half of the input amplitude thus causing a loss in output power of 6dB power.

I am uncertain if you agree with the above but I hope you do and your concern is simply with the scaling versus overflow. In relation to overflow I assume you are talking about overflow at the output and that the filter can handle internal overflow (2s complement wraparound) ?

If yes I do not understand your statement that `This might be fine for the DVB-S2 case but not for all signals which auto_baseband_rate tries to be.`  The existing `fir_128_4`  does not seem to be scaled to avoid overflow for all signals ? At least if my assumptions are correct the maximum output of the filter can overflow by a factor of 2.406 compared to the maximum possible value. Similar for the `fir_128_2` filter the maximum output of the filter can overflow by a factor of 2.767. 

If you are saying that overflow cannot occur in the existing filters I have clearly misunderstood something fundamental and would appreciate if you had the time to comment on my assumptions and elaborate on the inner workings.

**Comment by tfcollins** [comment]
> * The TX output power should not be affected by using the FIR filter provided by this lib right ?
> * Or, at the very least, it should be consistent across samples-rates so we do not observe a sudden 6 dB drop when going from 21 MHz to 19 MHz sample rate.

These are not the outright goals of these filters. The purpose of the filters are really three things:
1. Enable faster clocks by increasing decimation/interpolation
2. Passband compensation
3. Allow samples rates below 2.08 MS/s.

But we also try to do these things without distorting the input signal when possible, which will have tradeoffs. Now if you want to use other filters with different purposes you can totally do this with custom filter files, which can be designed with this tool: https://wiki.analog.com/resources/eval/user-guides/ad-fmcomms2-ebz/software/filters?s[]=filter&s[]=wizard , or by hand if you really want to. That filter wizard will even provide backoff requirements based on the filter you design.

> * The output of the filter is truncated back 16 or 12 bit width. For simplicity lets assume 16 bit. This means the 15 LSBs (and a redundant sign bit) are removed at the filter output since that is 32 bit after multiplication.

The inputs and outputs of all stages are not 16-bits. 

> I am uncertain if you agree with the above but I hope you do and your concern is simply with the scaling versus overflow. 

I am not certain this is necessarily the case. I would verify with the transceiver support group here to verify how gain is applied based on int 2 verse int 4: https://ez.analog.com/wide-band-rf-transceivers/design-support

> In relation to overflow I assume you are talking about overflow at the output and that the filter can handle intermediate, internal overflow (2s complement wraparound) ?

This is my direct concern, but it's not that the HW cannot handle intermediate bit expansion but since your filter has gain. So if I put in a full-scale signal it will saturate.

> If you are saying that overflow cannot occur in the existing filters I have clearly misunderstood something fundamental and would appreciate if you had the time to comment on my assumptions and elaborate on the inner workings.

When I simply compare a full-scale input between the existing code and your PR, the existing code does not saturate but your PR does. This might not be the case for your waveforms depending on its PAPR.

If you want to dig deeper into the datapath I would post in our support forums: https://ez.analog.com/wide-band-rf-transceivers/design-support

**Comment by Wramberg** [comment]
Thanks @tfcollins for the detailed reply. I am using the filters for passband compensation and was wrongly assuming/expecting that they also compensate for interpolation amplitude loss.

I realize this is not the case and although I must admit I find the behavior odd I guess there are reason which I have yet to understand. I will post on the support forum to get more details. Unfortunately I have abandoned Matlab a long time ago in favor of Python so your filter tool is not that useful to me :-) (I will have a look at the source code though)

> This is my direct concern, but it's not that the HW cannot handle intermediate bit expansion but since your filter has gain. So if I put in a full-scale signal it will saturate.

> When I simply compare a full-scale input between the existing code and your PR, the existing code does not saturate but your PR does. This might not be the case for your waveforms depending on its PAPR.

I still do not quite understand these statements. How does an interpolation factor of 2 coupled with the `fir_128_2` filter not cause any drop in output power ? To me the `fir_128_2` must have a gain of 2 (decimal) that is amplitude compensating for the interpolation. So I am under the impression that I was proposing to increase the gain - not introduce gain. This will of course  increase risk of overflow but again it does not introduce it - I fail to see how the current filters cannot overflow.

In any case I will check out the support forum and look into designing my own filters. I thought the filters were supposed to compensate for amplitude loss in interpolation but I was mistaken. Thanks for helping me clear it up.

**Pull request closed without merging**